### PR TITLE
[DRAFT] [libc++] [P1222] std::flat_set

### PR DESCRIPTION
**Pull request by Quuxplusone**
This is just a draft, but I'm putting it up in case people have early feedback.

TODO before this leaves the draft stage:
- Tests for all the constructors and member functions involving `sorted_unique`
- [LWG3803 "flat_foo constructors taking KeyContainer lack KeyCompare parameter"](https://cplusplus.github.io/LWG/issue3803)
- [LWG3804 "flat_foo missing some allocator-extended deduction guides"](https://cplusplus.github.io/LWG/issue3804)
- `_LIBCPP_HIDE_FROM_ABI` and `_LIBCPP_TEMPLATE_VIS` markup (I'm delaying this to keep the diff more readable)

Then after that:
- Should `flat_multiset` be done in the same file, or should we granularize this into `<__flat_set/flat_set.h>` and `<__flat_set/flat_multiset.h>` before taking the next step? I tentatively think "two files," but only because that would make it easier to ensure things like "the identifier `flat_set` doesn't accidentally appear anywhere in `flat_multiset.h` nor vice versa." We don't expect any future standard header ever to include either of the granular headers alone.
- *If* we introduce `__flat_set/`, should `sorted_unique.h` and `sorted_equivalent.h` move into that directory? I tentatively think "yes," but I wouldn't like to bet on these types' never being reused by another STL facility. Compare `__utility/in_place.h` (used only in `<any>`, `<expected>`, `<optional>`, `<ranges>`, `<variant>`) but `__variant/monostate.h` (used only in `<variant>`).

### CODE REVIEW

File: libcxx/include/flat_set [hunk 1]
```diff
@@ -0,0 +1,913 @@
...
+    : __c_(std::__make_obj_using_allocator<container_type>(__a, __cont)) {
+    auto __begin = __c_.begin();
+    auto __end = __c_.end();
+    std::stable_sort(__begin, __end, __comp_ref());
+    __c_.erase(std::__flatset_unique(__begin, __end, __comp_ref()), __end);
+  }
+
+  template <class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(container_type&& __cont, const _Alloc& __a)
```
Comment (ldionne) [inline libcxx/include/flat_set:169]: Please `#if 0` this until the LWG issue is discussed and agreed upon, with a comment about the LWG issue.

File: libcxx/include/flat_set [hunk 2]
```diff
@@ -0,0 +1,913 @@
...
+    : __compare_(__comp), __c_(std::__make_obj_using_allocator<container_type>(__a, __cont)) {
+    auto __begin = __c_.begin();
+    auto __end = __c_.end();
+    std::stable_sort(__begin, __end, __comp_ref());
+    __c_.erase(std::__flatset_unique(__begin, __end, __comp_ref()), __end);
+  }
+
+  template <class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(container_type&& __cont, const key_compare& __comp, const _Alloc& __a)
```
Comment (ldionne) [inline libcxx/include/flat_set:197]: Same, pending LWG issue.

File: libcxx/include/flat_set [hunk 3]
```diff
@@ -0,0 +1,913 @@
...
+  template <class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(sorted_unique_t, container_type&& __cont, const key_compare& __comp, const _Alloc& __a)
+    : __compare_(__comp), __c_(std::__make_obj_using_allocator<container_type>(__a, std::move(__cont))) {}
+
+  flat_set(const flat_set& __fs) = default;
+
+  flat_set(flat_set&& __fs)
+    noexcept(is_nothrow_constructible_v<_Compare, _Compare&> && is_nothrow_move_constructible_v<_KeyContainer>)
+    : __compare_(__fs.__compare_), __c_(std::move(__fs.__c_)) {
```
Comment (ldionne) [inline libcxx/include/flat_set:232]: Why are you not moving the comparator? You just explained it to me, but please record it on your list of issues.

File: libcxx/include/flat_set [hunk 4]
```diff
@@ -0,0 +1,913 @@
...
+  template <__container_compatible_range<value_type> _Rp, class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(from_range_t, _Rp&& __r, const key_compare& __comp, const _Alloc& __a)
+    : flat_set(__comp, __a) {
+    insert_range(std::forward<_Rp>(__r));
+  }
+
+  template <class _InIt>
+  flat_set(sorted_unique_t, _InIt __first, _InIt __last, const key_compare& __comp = key_compare())
+    : __compare_(__comp), __c_(__first, __last) {}
```
Comment (ldionne) [inline libcxx/include/flat_set:306]: I think that would be a nice place to add a `_LIBCPP_DEBUG_ASSERT(is_sorted)`. In fact, please make sure that you have a `_LIBCPP_ASSERT` for each precondition in this class. When the preconditions are expensive to check, we'll hide them behind the debug mode.

Comment (ldionne) [reply libcxx/include/flat_set:306]: Oh, and let's add a libc++ specific test for all of those, they are easy to write.

File: libcxx/include/flat_set [hunk 5]
```diff
@@ -0,0 +1,913 @@
...
+  flat_set(sorted_unique_t __s, initializer_list<value_type> __il,
+           const type_identity_t<key_compare>& __comp, const _Alloc& __a)
+    : flat_set(__s, __il.begin(), __il.end(), __comp, __a) {}
+
+  template <class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(sorted_unique_t __s, initializer_list<value_type> __il, const _Alloc& __a)
+    : flat_set(__s, __il.begin(), __il.end(), __a) { }
+
+  flat_set& operator=(initializer_list<value_type> __il) {
```
Comment (ldionne) [inline libcxx/include/flat_set:346]: Note to self: I reviewed all the constructors and stopped here.

**Review by ldionne — CHANGES_REQUESTED** [review]

File: libcxx/test/std/containers/container.adaptors/flat.set/allocator_mismatch.compile.fail.cpp [hunk 6]
```diff
@@ -0,0 +1,21 @@
+//===----------------------------------------------------------------------===//
```
Comment (ldionne) [inline libcxx/test/std/containers/container.adaptors/flat.set/allocator_mismatch.compile.fail.cpp:1]: This test (and the corresponding `std::set` test) should be a `.verify.cpp` test.

File: libcxx/include/flat_set [hunk 7]
```diff
@@ -0,0 +1,913 @@
...
+  flat_set& operator=(const flat_set& __fs) {
+    auto __guard = std::__make_exception_guard([&] { __restore_invariant(); });
+    __compare_ = __fs.__compare_;
+    __c_ = __fs.__c_;
+    __guard.__complete();
+    return *this;
+  }
+
+  flat_set& operator=(flat_set&& __fs)
+    noexcept(is_nothrow_assignable_v<_Compare&, _Compare&> && is_nothrow_assignable_v<_KeyContainer, _KeyContainer&&>)
```
Comment (ldionne) [inline libcxx/include/flat_set:361]: Should be `_KeyContainer&` -- also there should be a test.

File: libcxx/include/flat_set [hunk 8]
```diff
@@ -0,0 +1,913 @@
...
+      __guard.__complete();
+      return {__it, true};
+    } else {
+      return {__it, false};
+    }
+  }
+
+  template <class... _Args>
+  iterator emplace_hint(const_iterator, _Args&&... __args) {
+    return emplace(std::forward<_Args>(__args)...).first;
```
Comment (ldionne) [inline libcxx/include/flat_set:411]: I think we can actually use the hint. If the user gives us an iterator to the lower bound, then we can avoid computing it ourselves and perform the insertion directly at that location. You might be able to implement `emplace` in terms of `emplace_hint` in that case.

File: libcxx/include/flat_set [hunk 9]
```diff
@@ -0,0 +1,913 @@
...
+      auto __guard = std::__make_exception_guard([&] { __restore_invariant(); });
+      __it = __c_.emplace(__it, std::forward<_K2>(__x));
+      __guard.__complete();
+      return {__it, true};
+    } else {
+      return {__it, false};
+    }
+  }
+
+  iterator insert(const_iterator, const value_type& __x) { return emplace(__x).first; }
```
Comment (ldionne) [inline libcxx/include/flat_set:432]: Here and 2x below we can use the hint.

File: libcxx/include/flat_set [hunk 10]
```diff
@@ -0,0 +1,913 @@
...
+    return emplace(std::forward<_Args>(__args)...).first;
+  }
+
+  pair<iterator, bool> insert(const value_type& __x) { return emplace(__x); }
+  pair<iterator, bool> insert(value_type&& __x) { return emplace(std::move(__x)); }
+
+  template <class _K2>
+    requires __is_transparent<_Compare, _K2>::value &&
+             is_constructible_v<value_type, _K2>
+  pair<iterator, bool> insert(_K2&& __x) {
```
Comment (ldionne) [inline libcxx/include/flat_set:420]: We have a precondition that `find(value_type(__x)) == find(__x)` -- we could consider adding this in debug mode.

File: libcxx/include/flat_set [hunk 11]
```diff
@@ -0,0 +1,913 @@
...
+  }
+
+  template <class _InIt>
+    requires __is_cpp17_input_iterator<_InIt>::value
+  void insert(_InIt __first, _InIt __last) {
+    auto __guard = std::__make_exception_guard([&] { __restore_invariant(); });
+    auto __it = __c_.insert(__c_.end(), __first, __last);
+    auto __end = __c_.end();
+    if (__it != __end) {
+      std::stable_sort(__it, __end, __comp_ref());
```
Comment (ldionne) [inline libcxx/include/flat_set:449]: All the places where you use `stable_sort` need the resolution on the issue we discussed yesterday.

File: libcxx/include/flat_set [hunk 12]
```diff
@@ -0,0 +1,913 @@
...
+      auto __begin = __c_.begin();
+      std::inplace_merge(__begin, __it, __end, __comp_ref());
```
Comment (ldionne) [inline libcxx/include/flat_set:451]: As you said, I think you can do better here: you can remove duplicates as you're doing `inplace_merge`. That would be something like a funky `inplace_set_union` where the right hand side is potentially a multiset. Maybe that's too messy and not worth doing.

File: libcxx/include/flat_set [hunk 13]
```diff
@@ -0,0 +1,913 @@
...
+      __c_.erase(std::__flatset_unique(__begin, __end, __comp_ref()), __end);
+    }
+    __guard.__complete();
+  }
+
+  template <class _InIt>
```
Comment (ldionne) [inline libcxx/include/flat_set:457]: Can we call those `_InputIterator` instead of `_InIt`?

File: libcxx/include/flat_set [hunk 14]
```diff
@@ -0,0 +1,913 @@
...
+      std::inplace_merge(__begin, __it, __end, __comp_ref());
+      __c_.erase(std::__flatset_unique(__begin, __end, __comp_ref()), __end);
+    }
+    __guard.__complete();
+  }
+
+  void insert(std::initializer_list<value_type> __il) { insert(__il.begin(), __il.end()); }
+  void insert(sorted_unique_t __s, std::initializer_list<value_type> __il) { insert(__s, __il.begin(), __il.end()); }
+
+  container_type extract() && {
```
Comment (ldionne) [inline libcxx/include/flat_set:493]: Note to self, I stopped here.

**Review by ldionne — CHANGES_REQUESTED** [review]

File: libcxx/include/flat_set (hunk 7 above)
Comment (Quuxplusone) [reply libcxx/include/flat_set:361]: Turns out that I _did_ have a test which I now think is sufficient. The trick is that `is_nothrow_assignable_v<_KeyContainer, _KeyContainer&&>` isn't always-false. Because it's a type trait, it secretly means `is_nothrow_assignable_v<_KeyContainer&&, _KeyContainer&&>`, which is true unless `_KeyContainer` went out of its way to lvalue-ref-qualify its `operator=`. So I've just fixed the typo and left "flat.set/flat.set.cons/move_assign_noexcept.pass.cpp" as it was.

File: libcxx/test/std/containers/container.adaptors/flat.set/allocator_mismatch.compile.fail.cpp (hunk 6 above)
Comment (Quuxplusone) [reply libcxx/test/std/containers/container.adaptors/flat.set/allocator_mismatch.compile.fail.cpp:1]: Updated all the existing containers' `allocator_mismatch` tests in #8; PTAL!
After doing that, I found that this new test suffered from the exact failure mode of `.fail.cpp` tests — it wasn't testing what I thought it was! The third argument to `set` is an allocator, but the third argument to `flat_set` is supposed to be a container; so of course this test failed to compile as written. We could add a new test for `flat_set<int, less<int>, vector<int, allocator<long>>>`; but that's already covered by `vector/allocator_mismatch.verify.cpp`, so actually I think there's no longer anything to test here for `flat_set`.

File: libcxx/include/flat_set (hunk 1 above)
Comment (Quuxplusone) [reply libcxx/include/flat_set:169]: This is https://wg21.link/lwg3802 , and the fix is already done as a separate commit with "LWG3802" in the subject line.
The code for LWG3802 is unconditionally missing prior to the fix, and unconditionally present after the fix.
The tests for LWG3802 are `#if 0`'ed out prior to the fix, and unconditionally enabled after the fix.

If libc++ doesn't want to merge the fix for LWG3802 until [some-unspecified-point-in-the-future], that's fine; you can just omit that specific git commit. But I'm going to leave each LWG-issue-fixing commit in this actual PR branch for now, because that makes it easier to review the intended code/tests. (Otherwise you'll just ask over and over "Why is this part of the code wrong?" and the answer will always be "Because it doesn't become correct until after LWG issue such-and-such." It doesn't make sense to review wrong code.)

File: libcxx/include/flat_set (hunk 2 above)
Comment (Quuxplusone) [reply libcxx/include/flat_set:197]: https://wg21.link/lwg3803 , fixed in a separate commit with "LWG3803" in the subject line.

File: libcxx/include/flat_set (hunk 3 above)
Comment (Quuxplusone) [reply libcxx/include/flat_set:232]: (For the GitHub record: We're copying the comparator because the comparator might be a `std::function<bool(int,int)>`, or a lambda capturing a `std::string`, or anything of that nature. Moving-out-of such a comparator would leave the right-hand `flat_set` in an _invalid_ state — where e.g. `rhs.insert(42)` would crash the program — instead of the _valid but unspecified_ state that is required for conformance.)

File: libcxx/include/flat_set (hunk 10 above)
Comment (Quuxplusone) [reply libcxx/include/flat_set:420]: Done; but notice that libc++ does not bother with that precondition on either [`set::find(K&&)`](https://github.com/llvm/llvm-project/blob/4f15267/libcxx/include/set#L847-L856) or [`unordered_set::find(K&&)`](https://github.com/llvm/llvm-project/blob/4f15267/libcxx/include/unordered_set#L1424-L1435).

File: libcxx/include/flat_set (hunk 8 above)
Comment (Quuxplusone) [reply libcxx/include/flat_set:411]: Done. I factored out an `__emplace_unchecked(pos, K2&& x)` that assumes the insertion _does_ happen, at the _exact_ location passed to it; all the insert/emplace sites use that now.

FWIW, I see that back in 2019 @tzlaine and I discussed this exact API (on the cpplang Slack) and Zach was at least briefly considering not to provide it at all. His reasoning was:
- `mySet.insert(x)` is O(lg n), but `mySet.insert(pos, x)` with the right hint is O(1).
- `myFlatSet.insert(x)` is O(n), but `myFlatSet.insert(pos, x)` with the right hint is... still O(n).
Does that change your calculus at all? (Of course at this point I've already done the implementation... but it's still a fair bit of code to maintain and test, and I admit I haven't written any good tests for all the `emplace_hint` corner cases yet. Ignoring the hint would be slightly less code and way fewer obscure branches to test.)

File: libcxx/test/libcxx/containers/container.adaptors/flat.set/version.pass.cpp [hunk 15]
```diff
@@ -0,0 +1,20 @@
+//===----------------------------------------------------------------------===//
```
Comment (ldionne) [inline libcxx/test/libcxx/containers/container.adaptors/flat.set/version.pass.cpp:1]: Let's make this a `.compile.pass.cpp` test.

**Review by ldionne — CHANGES_REQUESTED** [review]

File: libcxx/include/flat_set [hunk 16]
```diff
@@ -0,0 +1,1306 @@
...
+    return 0;
+  }
+
+  template <class _K2>
+    requires __is_transparent<_Compare, _K2>::value &&
+             (!is_convertible_v<_K2&&, iterator>)
+  _LIBCPP_HIDE_FROM_ABI
+  size_type erase(_K2&& __x) {
+    // Notice that our iterator and const_iterator are the same type
+    auto [__first, __last] = equal_range(__x);
```
[the remark this answers is not in the document]
Comment (Quuxplusone) [reply libcxx/include/flat_set:862]: Yes. Done.

File: libcxx/include/flat_set [hunk 17]
```diff
@@ -0,0 +1,1306 @@
...
+
+  // observers
+
+  _LIBCPP_HIDE_FROM_ABI key_compare key_comp() const { return __compare_; }
+  _LIBCPP_HIDE_FROM_ABI value_compare value_comp() const { return __compare_; }
+
+  // set operations
+
+  _LIBCPP_HIDE_FROM_ABI
+  iterator find(const key_type& __x) {
```
[the remark this answers is not in the document]
Comment (Quuxplusone) [reply libcxx/include/flat_set:898]: I don't mind doing this, as long as `set` and `unordered_set` are brought into line too. This is now #10. (I'm not sure what sort of comment to leave in the code; let's take that discussion to #10.)

File: libcxx/include/flat_set [hunk 18]
```diff
@@ -0,0 +1,1306 @@
...
+    requires __is_transparent<_Compare, _K2>::value
+  _LIBCPP_HIDE_FROM_ABI
+  bool contains(const _K2& __x) const {
+    return find(__x) != end();
+  }
+
+  _LIBCPP_HIDE_FROM_ABI
+  iterator lower_bound(const key_type& __x) {
+    return std::partition_point(begin(), end(), [&](const value_type& __e) {
+      return bool(__compare_(__e, __x));
```
[the remark this answers is not in the document]
Comment (Quuxplusone) [reply libcxx/include/flat_set:965]: https://godbolt.org/z/ssejzdeh9
libc++ `set` is OK with explicit conversion to bool. libc++ `map` is not.
MSVC `set` is OK, but `map` is not (at least, with the old MSVC currently on Godbolt; the MSVC machines are offline right now).
libstdc++ `set` and `map` are not OK.

Therefore I think it's reasonable to remove all the casts to bool throughout. Will do.

File: libcxx/test/libcxx/containers/container.adaptors/flat.set/version.pass.cpp (hunk 15 above)
Comment (Quuxplusone) [reply libcxx/test/libcxx/containers/container.adaptors/flat.set/version.pass.cpp:1]: All the existing `version.pass.cpp` tests are not compile tests (e.g. `libcxx/test/libcxx/containers/sequences/vector/version.pass.cpp`). Do you want me to make that patch, or will you?

Comment (Quuxplusone) [reply libcxx/test/libcxx/containers/container.adaptors/flat.set/version.pass.cpp:1]: On a similar subject, #11.
